aptos-move/framework/aptos-framework/sources/Voting.move

```diff
@@ -0,0 +1,277 @@
+ * anyone can call resolve which returns the content of the proposal (of type ProposalType) that can be used to execute.
+ * 7. To ensure that a proposal can only be executed in specific ways, the governce can - (1) Have ProposalType include
+ * extra data so it can only be used in specific ways (enforced by where the change happens) or (2) pass the hash of the
+ * script module's bytecode when creating a proposal. When Vote::resolve is called later, it verifies that it's being
+ * called as part of a script module transaction whose hash matches the record.
+ */
+module AptosFramework::Voting {
+    use Std::Errors;
+    use Std::Event::{Self, EventHandle};
+    use Std::Option::{Self, Option};
+    use Std::Signer;
+    use Std::Vector;
+
+    use AptosFramework::Table::{Self, Table};
+    use AptosFramework::Timestamp;
+    use AptosFramework::TypeInfo::{Self, TypeInfo};
+
+    /// Error codes.
+    const EPROPOSAL_EXECUTION_HASH_NOT_MATCHING: u64 = 1;
+    const EPROPOSAL_CANNOT_BE_RESOLVED: u64 = 2;
+
+    /// ProposalStateEnum representing proposal state.
+    const PROPOSAL_STATE_PENDING: u64 = 0;
+    const PROPOSAL_STATE_SUCCEEDED: u64 = 1;
+    /// Proposal has failed because either the min vote threshold is not met or majority voted no.
+    const PROPOSAL_STATE_FAILED: u64 = 2;
+
+    struct Proposal<ProposalType: store> has store {
+        /// Should contain enough information to execute later, for example the required capability.
+        /// This is stored as a vector so we can return it to governance when the proposal is resolved.
+        execution_content: vector<ProposalType>,
+        /// (Optional) The hash for the execution script module. If present, only the same exact script module can
+        /// resolve this proposal.
+        execution_hash: Option<vector<u8>>,
+        /// A proposal is only resolved if expiration has passed and the number of votes is above threshold.
+        min_vote_threshold: u64,
+        expiration_secs: u64,
+
+        /// Number of votes for each outcome.
+        yes_votes: u64,
+        no_votes: u64,
+
+        /// Whether the proposal has already been resolved. This is to prevent double resolution.
+        resolved: bool,
+    }
+
+    struct VotingForum<ProposalType: store> has key {
+        /// Use Table for execution optimization instead of Vector for gas cost since Vector is read entirely into memory
+        /// during execution while only relevant Table entries are.
+        proposals: Table<u64, Proposal<ProposalType>>,
+        events: VotingEvents,
+        /// Unique identifier for a proposal. This allows for 2 * 10**19 proposals.
+        next_proposal_id: u64,
+    }
+
+    struct VotingEvents has store {
+        create_proposal_events: EventHandle<CreateProposalEvent>,
+        register_forum_events: EventHandle<RegisterForumEvent>,
+        vote_events: EventHandle<VoteEvent>,
```

Comment:
it's more intuitive to me to link the vote events in proposal instead of here

Reply:
If we store vote events inside each proposal, it can make it hard for off-chain to easily find vote events for all proposals in a voting forum/DAO. I think it's more convenient being outside like this.